File: incre-tests/synduce/combine/mts_and_mps.f
Inductive List = nil Unit | cons {Int, List};

max = \a: Int. \b: Int. if < a b then b else a;

mts = fix (
  \f: Int -> List -> Int. \pre: Int. \xs: List.
  match xs with
    nil _ -> pre
  | cons {h, t} -> f (max 0 (+ h pre)) t
  end
);

mps = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> max (+ h (f t)) 0
  end
);

spec = \xs: List. {mts 0 xs, mps xs};

repr = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> cons {h, f t}
  end
);

main = \xs: List. spec (repr xs);

File: ml-benchmark/synduce/combine/mts_and_mps.ml
config EnableDeepCoder = true

type 'a list =
  | Nil
  | Cons of 'a * 'a list

let rec sum = function
  | Nil -> 0
  | Cons (hd, tl) -> hd + sum tl

let max a b = if a > b then a else b

let rec mts s = function 
  | Nil -> s
  | Cons (hd, tl) -> mts (max (s + hd) 0) tl

let rec mps = function
  | Nil -> 0
  | Cons (hd, tl) -> max (mps tl + hd) 0

let spec l = (mts 0 l, mps l)

val repr: int list -> (int list) compress
let rec repr = function 
  | Nil -> Nil 
  | Cons (h, t) -> Cons (h, repr t)

let prog xs = spec (repr xs)